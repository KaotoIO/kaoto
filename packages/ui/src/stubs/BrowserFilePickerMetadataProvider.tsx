import { ChangeEvent, createRef, FunctionComponent, PropsWithChildren, useCallback, useRef } from 'react';
import {
  SCHEMA_FILE_ACCEPT_PATTERN,
  SCHEMA_FILE_ACCEPT_PATTERN_SOURCE_BODY,
  SCHEMA_FILE_NAME_PATTERN,
  SCHEMA_FILE_NAME_PATTERN_SOURCE_BODY,
} from '../models/datamapper';
import { IMetadataApi, MetadataContext } from '../providers';
import { readFileAsString } from './read-file-as-string';

export const BrowserFilePickerMetadataProvider: FunctionComponent<PropsWithChildren> = (props) => {
  const fileInputRef = createRef<HTMLInputElement>();
  const fileSelectionRef = useRef<{
    resolve: (files: Record<string, string>) => void;
    reject: (error: unknown) => unknown;
  }>();
  const fileContentsRef = useRef<Record<string, string>>();

  const askUserForFileSelection = useCallback(
    (
      include: string,
      _exclude?: string,
      _options?: Record<string, unknown>,
    ): Promise<string[] | string | undefined> => {
      if (!fileInputRef.current) return Promise.resolve(undefined);

      if (include === SCHEMA_FILE_NAME_PATTERN_SOURCE_BODY) {
        fileInputRef.current.accept = SCHEMA_FILE_ACCEPT_PATTERN_SOURCE_BODY;
      } else if (include === SCHEMA_FILE_NAME_PATTERN) {
        fileInputRef.current.accept = SCHEMA_FILE_ACCEPT_PATTERN;
      }

      fileInputRef.current.click();
      return new Promise<Record<string, string>>((resolve, reject) => {
        fileSelectionRef.current = { resolve, reject };
      }).then((files) => {
        fileContentsRef.current = files;
        return Object.keys(files);
      });
    },
    [fileInputRef],
  );

  const onImport = useCallback(async (event: ChangeEvent<HTMLInputElement>) => {
    const schemaFiles = event.target.files;
    if (!schemaFiles) return;
    const fileContents: Record<string, string> = {};
    const fileContentPromises: Promise<string>[] = [];
    Array.from(schemaFiles).forEach((f) => {
      const promise = readFileAsString(f).then((content) => (fileContents[f.name] = content));
      fileContentPromises.push(promise);
    });
    await Promise.allSettled(fileContentPromises);

    fileSelectionRef.current?.resolve(fileContents);
    fileSelectionRef.current = undefined;
    event.target.value = '';
  }, []);

  const getResourceContent = useCallback((path: string) => {
    return Promise.resolve(fileContentsRef.current && fileContentsRef.current[path]);
  }, []);

  const metadataApi: IMetadataApi = {
    askUserForFileSelection: askUserForFileSelection,
    getResourceContent: getResourceContent,
    shouldSaveSchema: true,
    getMetadata: () => Promise.resolve(undefined),
    setMetadata: () => Promise.resolve(),
    deleteResource: () => Promise.resolve(true),
    saveResourceContent: () => Promise.resolve(),
    getSuggestions: () => Promise.resolve([]),
    onStepUpdated: () => Promise.resolve(),
  };

  return (
    <MetadataContext.Provider value={metadataApi}>
      {props.children}
      <input
        type="file"
        style={{ display: 'none' }}
        data-testid={`attach-schema-file-input`}
        onChange={onImport}
        ref={fileInputRef}
      />
    </MetadataContext.Provider>
  );
};
