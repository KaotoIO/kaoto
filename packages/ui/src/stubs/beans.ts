export const beansYaml = `
- beans:
  - name: myBean
    type: io.kaoto.MyBean
    properties:
      p1: p1v
      p2: p2v
  - name: myBean2
    type: io.kaoto.MyBean
    properties:
      p1:
        p1s1: p1s1v
`

export const beansJson = {
  "beans": [
    {
      "name": "myBean",
      "type": "io.kaoto.MyBean",
      "properties": {
        "p1": "p1v",
        "p2": "p2v"
      }
    },
    {
      "name": "myBean2",
      "type": "io.kaoto.MyBean",
      "properties": {
        "p1": {
          "p1s1": "p1s1v"
        }
      }
    }
  ]
}

