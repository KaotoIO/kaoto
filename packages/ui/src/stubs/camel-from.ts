import { FromDefinition } from '@kaoto-next/camel-catalog/types';
import { CamelRouteVisualEntity } from '../models/visualization/flows';

/**
 * This is a stub Camel From in YAML format.
 * It is used to test the Canvas component.
 */
export const camelFromYaml = `
- from:
    uri: timer:tutorial
    steps:
    - to:
        uri: direct:my-route
`;

/**
 * This is a stub Camel From in JSON format.
 * It is used to test the Canvas component.
 */
export const camelFromJson: { from: FromDefinition } = {
  from: {
    uri: 'timer:tutorial',
    steps: [
      {
        to: {
          uri: 'direct:my-route',
        },
      },
    ],
  },
};

export const camelFrom = new CamelRouteVisualEntity({ from: camelFromJson.from });
