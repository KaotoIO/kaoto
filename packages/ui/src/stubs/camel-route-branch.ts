import { parse } from 'yaml';

export const camelRouteBranch = parse(`
- route:
    id: route-2768
    from:
      id: from-4014
      uri: timer:template
      parameters:
        period: "1000"
      steps:
        - choice:
            id: choice-3431
            otherwise:
              id: otherwise-3653
              steps:
                - log:
                    id: log-6808
                    message: \${body}
            when:
              - id: when-4112
                steps:
                  - setHeader:
                      id: setHeader-6078
                      expression:
                        simple: {}
                expression:
                  simple:
                    expression: \${header.foo} == 1
        - to:
            id: to-3757
            uri: sql
            parameters: {}
`);
