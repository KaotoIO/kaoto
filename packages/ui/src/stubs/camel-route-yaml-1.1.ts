export const camelRouteYaml_1_1_original = `
- route:
    id: route-3376
    from:
      id: from-3505
      uri: rest:post:/newCustomer
      parameters:
        host: localhost
      steps: []
`;

export const camelRouteYaml_1_1_updated = `
- route:
    id: route-3376
    from:
      id: from-3505
      uri: rest:post:/newCustomer
      parameters:
        bindingMode: "off"
        host: localhost
      steps: []
`;
