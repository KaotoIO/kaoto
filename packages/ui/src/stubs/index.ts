export * from './camel-route';
export * from './kamelet-binding-route';
export * from './kamelet-route';
export * from './pipe';
