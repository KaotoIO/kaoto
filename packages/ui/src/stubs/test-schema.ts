import { KaotoSchemaDefinition } from '../models/kaoto-schema';

/**
 * This is not a real schema, this is a hybrid custom made schema.
 */
export const testSchema: KaotoSchemaDefinition['schema'] = {
  type: 'object',
  definitions: {
    testRef: {
      type: 'object',
      title: 'testRef',
      properties: {
        spec: {
          type: 'string',
          title: 'Specification',
          description: 'Path to the OpenApi specification file.',
        },
      },
      required: ['spec'],
    },
  },
  properties: {
    id: {
      title: 'Id',
      type: 'string',
    },
    description: {
      title: 'Description',
      type: 'string',
    },
    uri: {
      title: 'Uri',
      type: 'string',
    },
    variableReceive: {
      title: 'Variable Receive',
      type: 'string',
    },
    testRef: {
      $ref: '#/definitions/testRef',
    },
    parameters: {
      type: 'object',
      title: 'Endpoint Properties',
      description: 'Endpoint properties description',
      properties: {
        timerName: {
          title: 'Timer Name',
          type: 'string',
        },
        delay: {
          title: 'Delay',
          type: 'string',
          default: '1000',
        },
        fixedRate: {
          title: 'Fixed Rate',
          type: 'boolean',
          default: false,
        },
        includeMetadata: {
          title: 'Include Metadata',
          type: 'boolean',
          default: false,
        },
        period: {
          title: 'Period',
          type: 'string',
          default: '1000',
        },
        repeatCount: {
          title: 'Repeat Count',
          type: 'integer',
        },
        exceptionHandler: {
          title: 'Exception Handler',
          type: 'string',
          $comment: 'class:org.apache.camel.spi.ExceptionHandler',
        },
        exchangePattern: {
          title: 'Exchange Pattern',
          type: 'string',
          enum: ['InOnly', 'InOut'],
        },
        synchronous: {
          title: 'Synchronous',
          type: 'boolean',
          default: false,
        },
        timer: {
          title: 'Timer',
          type: 'string',
          $comment: 'class:java.util.Timer',
        },
        runLoggingLevel: {
          title: 'Run Logging Level',
          type: 'string',
          default: 'TRACE',
          enum: ['TRACE', 'DEBUG', 'INFO', 'WARN', 'ERROR', 'OFF'],
        },
      },
      required: ['timerName', 'fixedRate', 'repeatCount', 'exceptionHandler', 'runLoggingLevel'],
    },
    kameletProperties: {
      title: 'Properties',
      type: 'array',
      description: 'Configure properties on the Kamelet',
      items: {
        type: 'object',
        properties: {
          name: {
            title: 'Property name',
            description: 'Name of the property',
            type: 'string',
          },
          title: {
            title: 'Title',
            description: 'Display name of the property',
            type: 'string',
          },
          description: {
            title: 'Description',
            description: 'Simple text description of the property',
            type: 'string',
          },
          type: {
            title: 'Property type',
            description: 'Set the expected type for this property',
            type: 'string',
            enum: ['string', 'number', 'boolean'],
            default: 'string',
          },
          default: {
            title: 'Default',
            description: 'Default value for the property',
            type: 'string',
          },
          'x-descriptors': {
            title: 'X-descriptors',
            description: 'Specific aids for the visual tools',
            type: 'array',
            items: {
              type: 'string',
            },
          },
        },
        required: ['name', 'type'],
      },
    },
    labels: {
      additionalProperties: {
        default: '',
        type: 'string',
      },
      title: 'Additional Labels',
      description:
        'Map of string keys and values that can be used to organize and categorize (scope and select) objects. May match selectors of replication controllers and services. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels',
      type: 'object',
    },
  },
  required: ['id', 'uri', 'labels'],
};

export const inputModelForTestSchema: Record<string, unknown> = {
  id: 'from-2860',
  uri: 'test',
  variableReceive: 'test',
  testRef: {
    spec: 'test',
  },
  parameters: {
    period: '1000',
    timerName: 'template',
    exceptionHandler: '#test',
    exchangePattern: 'InOnly',
    fixedRate: true,
    runLoggingLevel: 'OFF',
    repeatCount: '2',
    time: undefined,
    timer: '#testbean',
  },
  kameletProperties: [
    {
      name: 'period',
      title: 'Period',
      description: 'The time interval between two events',
      type: 'integer',
      default: 5000,
    },
  ],
  labels: {
    test: 'test',
  },
  steps: [
    {
      log: {
        id: 'log-2942',
        message: 'template message',
      },
    },
  ],
};
