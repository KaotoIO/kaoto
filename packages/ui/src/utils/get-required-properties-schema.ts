import { KaotoSchemaDefinition } from '../models';
import { isDefined } from './is-defined';
import { resolveRefIfNeeded } from './resolve-ref-if-needed';

/**
 * Extracts a schema containing only the required properties.
 * Recursively resolves `$ref` if necessary.
 */
export function getRequiredPropertiesSchema(
  schema?: KaotoSchemaDefinition['schema'],
  resolveFromSchema?: KaotoSchemaDefinition['schema'],
): KaotoSchemaDefinition['schema'] {
  if (!isDefined(schema) || !isDefined(resolveFromSchema)) return {};

  const schemaProperties = schema.properties;
  const requiredProperties = schema.required as string[];

  if (!isDefined(schemaProperties)) {
    return { ...schema, properties: {} };
  }

  const requiredFormSchema = Object.entries(schemaProperties).reduce(
    (acc, [property, definition]) => {
      if ('$ref' in definition) {
        const objectDefinition = resolveRefIfNeeded(definition, resolveFromSchema);
        const subSchema = getRequiredPropertiesSchema(objectDefinition, resolveFromSchema);
        if (Object.keys(subSchema.properties as object).length > 0) {
          acc[property] = subSchema;
        }
      } else if (definition['type'] === 'object' && 'properties' in definition) {
        const subSchema = getRequiredPropertiesSchema(definition, resolveFromSchema);
        if (Object.keys(subSchema.properties as object).length > 0) {
          acc[property] = subSchema;
        }
      } else if (isDefined(requiredProperties) && requiredProperties.indexOf(property) > -1) {
        acc[property] = definition;
      }

      return acc;
    },
    {} as KaotoSchemaDefinition['schema'],
  );

  return { ...schema, properties: requiredFormSchema };
}
