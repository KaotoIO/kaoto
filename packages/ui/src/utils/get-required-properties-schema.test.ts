import { getRequiredPropertiesSchema } from './get-required-properties-schema';
import { testSchema } from '../stubs/test-schema';

describe('getRequiredPropertiesSchema()', () => {
  it('should return only the properties which are Required', () => {
    const procesedSchema = getRequiredPropertiesSchema(testSchema, testSchema);
    expect(procesedSchema).toMatchSnapshot();
  });

  it('should return {}', () => {
    const procesedSchema = getRequiredPropertiesSchema({}, testSchema);
    expect(procesedSchema).toMatchObject({});
  });
});
