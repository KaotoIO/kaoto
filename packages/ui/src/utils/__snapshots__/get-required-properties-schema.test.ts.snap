// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRequiredPropertiesSchema() should return only the properties which are Required 1`] = `
{
  "definitions": {
    "testRef": {
      "properties": {
        "spec": {
          "description": "Path to the OpenApi specification file.",
          "title": "Specification",
          "type": "string",
        },
      },
      "required": [
        "spec",
      ],
      "title": "testRef",
      "type": "object",
    },
  },
  "properties": {
    "id": {
      "title": "Id",
      "type": "string",
    },
    "labels": {
      "additionalProperties": {
        "default": "",
        "type": "string",
      },
      "description": "Map of string keys and values that can be used to organize and categorize (scope and select) objects. May match selectors of replication controllers and services. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels",
      "title": "Additional Labels",
      "type": "object",
    },
    "parameters": {
      "description": "Endpoint properties description",
      "properties": {
        "exceptionHandler": {
          "$comment": "class:org.apache.camel.spi.ExceptionHandler",
          "title": "Exception Handler",
          "type": "string",
        },
        "fixedRate": {
          "default": false,
          "title": "Fixed Rate",
          "type": "boolean",
        },
        "repeatCount": {
          "title": "Repeat Count",
          "type": "integer",
        },
        "runLoggingLevel": {
          "default": "TRACE",
          "enum": [
            "TRACE",
            "DEBUG",
            "INFO",
            "WARN",
            "ERROR",
            "OFF",
          ],
          "title": "Run Logging Level",
          "type": "string",
        },
        "timerName": {
          "title": "Timer Name",
          "type": "string",
        },
      },
      "required": [
        "timerName",
        "fixedRate",
        "repeatCount",
        "exceptionHandler",
        "runLoggingLevel",
      ],
      "title": "Endpoint Properties",
      "type": "object",
    },
    "testRef": {
      "properties": {
        "spec": {
          "description": "Path to the OpenApi specification file.",
          "title": "Specification",
          "type": "string",
        },
      },
      "required": [
        "spec",
      ],
      "title": "testRef",
      "type": "object",
    },
    "uri": {
      "title": "Uri",
      "type": "string",
    },
  },
  "required": [
    "id",
    "uri",
    "labels",
  ],
  "type": "object",
}
`;
